src/measure_theory/integral_eq_improper.lean

```diff
@@ -0,0 +1,479 @@
+
+lemma ae_cover.inter_restrict {φ : ι → set α} (hφ : ae_cover μ l φ)
+  {s : set α} (hs : measurable_set s) :
+  ae_cover (μ.restrict s) l (λ i, φ i ∩ s) :=
+ae_cover_restrict_of_ae_imp hs
+  (hφ.ae_eventually_mem.mono (λ x hx hxs, hx.mono $ λ i hi, ⟨hi, hxs⟩))
+  (λ i, (hφ.measurable i).inter hs)
+
+lemma ae_cover.ae_tendsto_indicator {β : Type*} [has_zero β] [topological_space β]
+  {f : α → β} {φ : ι → set α} (hφ : ae_cover μ l φ) :
+  ∀ᵐ x ∂μ, tendsto (λ i, (φ i).indicator f x) l (𝓝 $ f x) :=
+hφ.ae_eventually_mem.mono (λ x hx, tendsto_const_nhds.congr' $
+  hx.mono $ λ n hn, (indicator_of_mem hn _).symm)
+
+end ae_cover
+
+lemma ae_cover.comp_tendsto {α ι ι' : Type*} [measurable_space α] {μ : measure α} {l : filter ι}
+  {l' : filter ι'} {φ : ι → set α} (hφ : ae_cover μ l φ)
+  {u : ι' → ι} (hu : tendsto u l' l) :
+  ae_cover μ l' (φ ∘ u) :=
+{ ae_eventually_mem := hφ.ae_eventually_mem.mono (λ x hx, hu.eventually hx),
+  measurable := λ i, hφ.measurable (u i) }
+
+section ae_cover_Union_Inter_encodable
+
+variables {α ι : Type*} [encodable ι]
+  [measurable_space α] {μ : measure α}
+
+lemma ae_cover.bUnion_Iic_ae_cover [preorder ι] {φ : ι → set α} (hφ : ae_cover μ at_top φ) :
+  ae_cover μ at_top (λ (n : ι), ⋃ k (h : k ∈ Iic n), φ k) :=
+{ ae_eventually_mem := hφ.ae_eventually_mem.mono
+    (λ x h, h.mono (λ i hi, mem_bUnion right_mem_Iic hi)),
+  measurable := λ i, measurable_set.bUnion (countable_encodable _) (λ n _, hφ.measurable n) }
+
+lemma ae_cover.bInter_Ici_ae_cover [semilattice_sup ι] [nonempty ι] {φ : ι → set α}
+  (hφ : ae_cover μ at_top φ) : ae_cover μ at_top (λ (n : ι), ⋂ k (h : k ∈ Ici n), φ k) :=
+{ ae_eventually_mem := hφ.ae_eventually_mem.mono
+    begin
+      intros x h,
+      rw eventually_at_top at *,
+      rcases h with ⟨i, hi⟩,
+      use i,
+      intros j hj,
+      exact mem_bInter (λ k hk, hi k (le_trans hj hk)),
+    end,
+  measurable := λ i, measurable_set.bInter (countable_encodable _) (λ n _, hφ.measurable n) }
+
+end ae_cover_Union_Inter_encodable
+
+section lintegral
+
+variables {α ι : Type*} [measurable_space α] {μ : measure α} {l : filter ι}
+
+lemma ae_cover.lintegral_tendsto_of_monotone_of_nat {φ : ℕ → set α} (hφ : ae_cover μ at_top φ)
+  (hmono : monotone φ) {f : α → ℝ≥0∞} (hfm : measurable f) :
+  tendsto (λ i, ∫⁻ x in φ i, f x ∂μ) at_top (𝓝 $ ∫⁻ x, f x ∂μ) :=
+let F := λ n, (φ n).indicator f in
+have key₁ : ∀ n, ae_measurable (F n) μ, from λ n, (hfm.indicator (hφ.measurable n)).ae_measurable,
+have key₂ : ∀ᵐ (x : α) ∂μ, monotone (λ n, F n x), from ae_of_all _
+  (λ x i j hij, indicator_le_indicator_of_subset (hmono hij) (λ x, zero_le $ f x) x),
+have key₃ : ∀ᵐ (x : α) ∂μ, tendsto (λ n, F n x) at_top (𝓝 (f x)), from hφ.ae_tendsto_indicator,
+(lintegral_tendsto_of_tendsto_of_monotone key₁ key₂ key₃).congr
+  (λ n, lintegral_indicator f (hφ.measurable n))
+
+lemma ae_cover.lintegral_tendsto_of_nat {φ : ℕ → set α} (hφ : ae_cover μ at_top φ)
+  {f : α → ℝ≥0∞} (hfm : measurable f) :
```

Comment:
in this lemma, you are assuming that `f` is measurable. However, I think you can easily deduce from this one a version assuming only that `f` is almost everywhere measurable. From there, you can weaken all the assumptions in the subsequent lemmas, from measurability to almost everywhere measurability.

Reply:
Instead of just using the `measurable` version to prove the `ae_measurable` one, I added a few lemmas to `integrable_on.lean` about ae-measurability of an indicator, which allows me to use `ae_measurable` all the way. Should I move these to a separate PR ?